README.md
# bulk-pack-transformer

API service that handles requests with payloads (from data packs uploaded from devices) and creates field data and other models used by Greenstand platform.


CHANGELOG.md
## [1.6.9](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.8...v1.6.9) (2023-05-30)


### Bug Fixes

* semantic release ([4e4f369](https://github.com/Greenstand/bulk-pack-transformer/commit/4e4f369ecbea7bca770f948c3e63add65c4991c8))
* uuid duplicate check ([14555a7](https://github.com/Greenstand/bulk-pack-transformer/commit/14555a7c5584ab46014a9cbafad8d009478d93d5))

## [1.6.8](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.7...v1.6.8) (2022-06-07)


### Bug Fixes

* rebuild package-lock.json ([392f79a](https://github.com/Greenstand/bulk-pack-transformer/commit/392f79a0025b66b63b1f13186e90892b22f60eae))

## [1.6.7](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.6...v1.6.7) (2022-06-07)


### Bug Fixes

* rebuild package-lock.json ([81d6c8f](https://github.com/Greenstand/bulk-pack-transformer/commit/81d6c8fb855a326f2bfb4ec37ba27891dd68cc86))

## [1.6.6](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.5...v1.6.6) (2022-06-07)


### Bug Fixes

* rebuild package-lock.json ([b0c62b6](https://github.com/Greenstand/bulk-pack-transformer/commit/b0c62b6be2411be2126ce7cb581fcf00bf92d2fc))

## [1.6.5](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.4...v1.6.5) (2022-06-06)


### Bug Fixes

* rotation_matrix ([5b2a340](https://github.com/Greenstand/bulk-pack-transformer/commit/5b2a340717de6b6d32ba908caf96ba3d95870761))

## [1.6.4](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.3...v1.6.4) (2022-03-06)


### Bug Fixes

* rotation_matrix ([1bbd087](https://github.com/Greenstand/bulk-pack-transformer/commit/1bbd08776eb98ca44c6bf52b7fca92ce633191f0))

## [1.6.3](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.2...v1.6.3) (2022-03-04)


### Bug Fixes

* change Math.round to Math.floor ([e9affcc](https://github.com/Greenstand/bulk-pack-transformer/commit/e9affccfa70c6e263d7eeeadb04781402d2fb272))
* round ios_sdk_version value to int ([c8352d1](https://github.com/Greenstand/bulk-pack-transformer/commit/c8352d170cd24fa2e2c02cc60d6123b14474d586))

## [1.6.2](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.1...v1.6.2) (2022-03-01)


### Bug Fixes

* add support for ios os properties ([00837e7](https://github.com/Greenstand/bulk-pack-transformer/commit/00837e7f6e42ed4a3afd05ea01f553cf55f8e5e5))

## [1.6.1](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.6.0...v1.6.1) (2022-02-28)


### Bug Fixes

* temporarily disable extra_attributes for v2 bulk pack ([35ceb9c](https://github.com/Greenstand/bulk-pack-transformer/commit/35ceb9c17bf65fb8254201ea642c11fe836f9624))

# [1.6.0](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.5.3...v1.6.0) (2022-02-27)


### Features

* rename capture_taken_at to captured_at for v2 ([97b3b30](https://github.com/Greenstand/bulk-pack-transformer/commit/97b3b3037b14437700ff95e5782505ac94456df6))

## [1.5.3](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.5.2...v1.5.3) (2022-02-22)


### Bug Fixes

* more logs ([765aa70](https://github.com/Greenstand/bulk-pack-transformer/commit/765aa70f57b0db0bf9c71dfbae1f7c4f8459a3ad))

## [1.5.2](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.5.1...v1.5.2) (2022-02-22)


### Bug Fixes

* debugging logs ([81b3b15](https://github.com/Greenstand/bulk-pack-transformer/commit/81b3b15d7f97ef42ca7a0f81d0cbf97aded78bde))
* Revert ([529a0da](https://github.com/Greenstand/bulk-pack-transformer/commit/529a0dad2e5e073c0d611ae6eb65f2f26d231ad3))

## [1.5.1](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.5.0...v1.5.1) (2022-02-22)


### Bug Fixes

* remove invalid attributes mapping ([f357d08](https://github.com/Greenstand/bulk-pack-transformer/commit/f357d08024130c28e7bb7629daafe179c9832042))

# [1.5.0](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.4.7...v1.5.0) (2022-02-15)


### Features

* bulk pack v2 updates ([46c93a3](https://github.com/Greenstand/bulk-pack-transformer/commit/46c93a3b391498e77206b57ab782da41420cc8eb))

## [1.4.7](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.4.6...v1.4.7) (2022-01-26)


### Bug Fixes

* correct namespace for sealed secret ([a4c2658](https://github.com/Greenstand/bulk-pack-transformer/commit/a4c2658083fc8ab61c64ee6fe08749f64c80992c))

## [1.4.6](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.4.5...v1.4.6) (2022-01-26)


### Bug Fixes

* correct ssl settings in db sealed secret ([696f77e](https://github.com/Greenstand/bulk-pack-transformer/commit/696f77ec13376c2d23f09eaac4af3626943f2426))

## [1.4.5](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.4.4...v1.4.5) (2021-12-31)


### Bug Fixes

* upgrade pg package ([9ad949f](https://github.com/Greenstand/bulk-pack-transformer/commit/9ad949f61503f11a4df62dd926af3ababb941eda))

## [1.4.4](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.4.3...v1.4.4) (2021-12-31)


### Bug Fixes

* add some more detailed debug messaging ([581e6bd](https://github.com/Greenstand/bulk-pack-transformer/commit/581e6bdc301162d19345f23d9ddfad99f9fda259))

## [1.4.3](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.4.2...v1.4.3) (2021-12-31)


### Bug Fixes

* update package-lock ([d2af3ba](https://github.com/Greenstand/bulk-pack-transformer/commit/d2af3ba1f75731da11905168d0b336f1ed2798ce))

## [1.4.2](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.4.1...v1.4.2) (2021-12-08)


### Bug Fixes

* reseal the secret with correct ssl ([8aabc31](https://github.com/Greenstand/bulk-pack-transformer/commit/8aabc31d0e3240ea3da20e0f4fc4eab3cdfe34b1))

## [1.4.1](https://github.com/Greenstand/bulk-pack-transformer/compare/v1.4.0...v1.4.1) (2021-12-02)


### Bug Fixes

* update treetracker database sealed secret ([0015a6a](https://github.com/Greenstand/bulk-pack-transformer/commit/0015a6a97a29b3880cc5e5ab762e2f0eb7b0145e))


index.spec.js
const supertest = require('supertest');
const sinon = require('sinon');
const chai = require('chai');
const sinonChai = require('sinon-chai');
chai.use(sinonChai);
const expect = chai.expect;
const { Pool } = require('pg');
const rp = require('request-promise-native');

const app = require('./index.js');
const request = supertest(app);

const conn = require('./config/config');
const Data = require('./src/data.js');
const pool = new Pool({
  connectionString: conn.connectionString,
});

after(async () => {
  // cleanups
  const getQuery = {
    text: `
      DELETE FROM trees WHERE true;
      DELETE FROM devices WHERE true;
      DELETE FROM planter_registrations WHERE true;
      DELETE FROM planter WHERE true;
    `,
  };
  await pool.query(getQuery).catch((err) => {
    console.log(`Database cleanup after running tests failed`);
    throw err;
  });
});

describe('API', () => {
  describe('Planters Endpoints', () => {
    describe('planter registrations', () => {
      let dataStub, data2Stub;

      beforeEach(() => {
        dataStub = sinon
          .stub(Data.prototype, 'findOrCreateUser')
          .resolves({ id: 1, phone: 'phone' });
        data2Stub = sinon
          .stub(Data.prototype, 'createPlanterRegistration')
          .resolves();
      });

      afterEach(() => {
        dataStub.restore();
        data2Stub.restore();
      });
      it('should register a planter at /planter', async () => {
        const v1PlanterObject = {
          planter_identifier: 'planter_identifier',
          first_name: 'first_name',
          last_name: 'last_name',
          organization: 'organization',
          device_identifier: 'device_identifier',
          record_uuid: 'record_uuid',
          image_url: 'image_url',
          lat: 'lat',
          lon: 'lon',
        };

        await request.post('/planter').send(v1PlanterObject).expect(200);

        const {
          planter_identifier,
          first_name,
          last_name,
          organization,
          device_identifier,
        } = v1PlanterObject;
        expect(dataStub).calledWith(
          planter_identifier,
          first_name,
          last_name,
          organization,
        );
        expect(data2Stub).calledWith(1, device_identifier, {
          ...v1PlanterObject,
          phone: 'phone',
          email: undefined,
        });
      });

      it('should register a planter at /v2/wallet_registrations', async () => {
        const v2PlanterObject = {
          wallet: 'wallet',
          first_name: 'first_name',
          last_name: 'last_name',
          phone: 'phone',
          email: 'email',
          lat: 'lat',
          lon: 'lon',
          user_photo_url: 'user_photo_url',
          registered_at: 'registered_at',
        };
        await request
          .post('/v2/wallet_registrations')
          .send(v2PlanterObject)
          .expect(200);

        const { wallet, first_name, last_name, lat, lon } = v2PlanterObject;

        expect(dataStub).calledWith(wallet, first_name, last_name, null);
        expect(data2Stub).calledWith(1, null, {
          planter_identifier: wallet,
          first_name,
          last_name,
          organization: null,
          device_identifier: null,
          lat,
          lon,
          phone: 'phone',
          email: undefined,
        });
      });
    });
  });

  describe('Devices Endpoints', () => {
    describe('PUT /devices/', () => {
      let dataStub;

      beforeEach(() => {
        dataStub = sinon
          .stub(Data.prototype, 'upsertDevice')
          .resolves({ status: 'Device upserted' });
      });

      afterEach(() => {
        dataStub.restore();
      });

      it('should add device to the list of devices at /devices', async () => {
        const deviceObject = {
          device_identifier: 'device_identifier',
          app_version: 'app_version',
          app_build: 'app_build',
          manufacturer: 'manufacturer',
          brand: 'brand',
          model: 'model',
          hardware: 'hardware',
          device: 'device',
          serial: 'serial',
          androidRelease: 'androidRelease',
          androidSdkVersion: 'androidSdkVersion',
        };

        const response = await request
          .put('/device')
          .send(deviceObject)
          .expect(200);

        expect(response.body).eql({ device: { status: 'Device upserted' } });
        expect(dataStub).calledWith(deviceObject);
      });

      it('should add device to the list of devices at /v2/device_configurations', async () => {
        const deviceObject = {
          device_identifier: 'device_identifier',
          brand: 'brand',
          model: 'model',
          device: 'device',
          serial: 'serial',
          hardware: 'hardware',
          manufacturer: 'manufacturer',
          app_build: 'app_build',
          app_version: 'app_version',
          os_version: 'os_version',
          sdk_version: 'sdk_version',
          logged_at: 'logged_at',
        };
        const response = await request
          .post('/v2/device_configurations')
          .send(deviceObject)
          .expect(200);

        expect(response.body).eql({ device: { status: 'Device upserted' } });
        const { os_version, sdk_version } = deviceObject;
        delete deviceObject.os_version;
        delete deviceObject.sdk_version;
        delete deviceObject.logged_at;
        expect(dataStub).calledWith({
          ...deviceObject,
          androidRelease: os_version,
          androidSdkVersion: sdk_version,
        });
      });
    });
  });

  describe('Trees Endpoints', () => {
    describe('POST /tree', () => {
      let findUserStub,
        checkForExistingTreeStub,
        createTreeStub,
        rpStub,
        useFieldDataServiceStub;

      beforeEach(() => {
        findUserStub = sinon.stub(Data.prototype, 'findUser');
        checkForExistingTreeStub = sinon.stub(
          Data.prototype,
          'checkForExistingTree',
        );
        createTreeStub = sinon.stub(Data.prototype, 'createTree');
        rpStub = sinon.stub(rp, 'post');
        useFieldDataServiceStub = sinon.stub(conn, 'useFieldDataService');
      });

      afterEach(() => {
        findUserStub.restore();
        checkForExistingTreeStub.restore();
        createTreeStub.restore();
        rpStub.restore();
        useFieldDataServiceStub.restore();
      });

      const treeObject = {
        uuid: 'uuid',
        image_url: 'image_url',
        lat: 'lat',
        lon: 'lon',
        gps_accuracy: 'gps_accuracy',
        note: 'note',
        device_identifier: 'device_identifier',
        planter_identifier: 'planter_identifier',
        timestamp: 1644620400000,
        attributes: [],
      };

      const v2TreeObject = {
        id: 'id',
        session_id: 'session_id',
        image_url: 'image_url',
        lat: 'lat',
        lon: 'lon',
        gps_accuracy: 'gps_accuracy',
        note: 'note',
        abs_step_count: 'abs_step_count',
        delta_step_count: 'delta_step_count',
        rotation_matrix: 'rotation_matrix',
        extra_attributes: 'extra_attributes',
        captured_at: '2022-02-12T22:45:36.005Z',
      };

      it('should error out -- planter identifier not found', async () => {
        const response = await request
          .post('/tree')
          .send(treeObject)
          .set('Accept', 'application/json')
          .expect(404);
        expect(findUserStub).calledWith('planter_identifier');
        expect(checkForExistingTreeStub).not.called;
        expect(createTreeStub).not.called;
        expect(response.body).eql({
          error: `planter not found planter_identifier`,
        });
      });

      it('duplicates exist return 200', async () => {
        findUserStub.resolves({ user: 'found' });
        checkForExistingTreeStub.resolves({ tree: 'found' });
        const response = await request
          .post('/tree')
          .send(treeObject)
          .set('Accept', 'application/json')
          .expect(200);
        expect(response.body).eql({
          duplicate: { tree: 'found' },
        });
        expect(findUserStub).calledWith('planter_identifier');
        expect(checkForExistingTreeStub).calledWith('uuid');
        expect(createTreeStub).not.called;
      });

      it('create a legacy tree -- useFieldDataService env variable is not true', async () => {
        findUserStub.resolves({ id: 'found' });
        createTreeStub.resolves({ tree: 'created' });
        checkForExistingTreeStub.resolves(null);
        useFieldDataServiceStub.get(() => 'false');
        const response = await request
          .post('/tree')
          .send(treeObject)
          .set('Accept', 'application/json')
          .expect(201);
        expect(response.body).eql({
          tree: { tree: 'created' },
        });
        expect(findUserStub).calledWith('planter_identifier');
        expect(rpStub).not.called;
        expect(checkForExistingTreeStub).calledWith('uuid');
        expect(createTreeStub).calledWith(
          'found',
          'device_identifier',
          treeObject,
        );
      });

      it('make a call to the field_data api', async () => {
        findUserStub.resolves({ id: 'found' });
        rpStub.resolves({ capture: 'created' });
        checkForExistingTreeStub.resolves(null);
        useFieldDataServiceStub.get(() => 'true');
        const response = await request
          .post('/tree')
          .send({ ...treeObject })
          .set('Accept', 'application/json')
          .expect(201);
        expect(response.body).eql({
          fieldCapture: { capture: 'created' },
        });
        expect(findUserStub).calledWith('planter_identifier');
        expect(rpStub).calledOnce;
        expect(checkForExistingTreeStub).calledWith('uuid');
        expect(createTreeStub).not.called;
      });

      it('make a call to the field_data api -- v2Captures', async () => {
        findUserStub.resolves({ id: 'found' });
        rpStub.resolves({ capture: 'created' });
        checkForExistingTreeStub.resolves(null);
        useFieldDataServiceStub.get(() => 'true');
        const response = await request
          .post('/v2/captures')
          .send(v2TreeObject)
          .set('Accept', 'application/json')
          .expect(201);
        expect(findUserStub).not.called;
        expect(checkForExistingTreeStub).calledWith('id');
        expect(createTreeStub).not.called;
        expect(response.body).eql({
          fieldCapture: { capture: 'created' },
        });
      });
    });
  });
});


src/data.spec.js
const Data = require('./data');
const { Pool } = require('pg');
const conn = require('./../config/config');
const pool = new Pool({
  connectionString: conn.connectionString,
});
const data = new Data(pool);

const expect = require('chai').expect;
const uuid = require('uuid');

after(async () => {
  // cleanups
  const getQuery = {
    text: `
      DELETE FROM trees WHERE true;
      DELETE FROM devices WHERE true;
      DELETE FROM planter_registrations WHERE true;
      DELETE FROM planter WHERE true;
      
    `,
  };
  await pool.query(getQuery).catch((err) => {
    console.log(`Database cleanup after running tests failed`);
    throw err;
  });
});

describe('The data functions', function () {
  it('creates a user', async () => {
    const createdUser = await data.findOrCreateUser(
      085739203636,
      'Test',
      'Test2',
      'Test3',
    );
    expect(createdUser).to.not.be.null;

    const existingUser = await data.findUser(085739203636);
    expect(existingUser).to.not.be.null;
  });

  it('creates a planter registration', async () => {
    const reg = await data.createPlanterRegistration(1, 11, {
      first_name: 'Test',
      last_name: 'User',
      organization: 'My Organization',
      planter_identifier: '123123123',
      lon: 12,
      lat: 12,
    });
    expect(reg).to.not.be.null;
  });

  it('creates a tree', async () => {
    const treeUUID = uuid();
    const tree = await data.createTree(1, 1, {
      lat: 80,
      lon: 120,
      gps_accuracy: 1,
      note: 'my note',
      timestamp: 1536367800,
      image_url: 'http://www.myimage.org/',
      sequence_id: 1,
      uuid: treeUUID,
    });
    expect(tree).to.not.be.null;

    const trees = await data.trees();
    expect(trees.length).to.equal(1);

    it('finds trees for a user', async () => {
      const trees = await data.treesForUser(1);
      expect(trees.length).to.equal(1);
    });
  });

  it('detects a duplicate tree', async () => {
    const treeUUID = uuid();
    const tree = await data.createTree(1, 1, {
      lat: 80,
      lon: 120,
      gps_accuracy: 1,
      note: 'my note',
      timestamp: 1536367800,
      image_url: 'http://www.myimage.org/',
      sequence_id: 1,
      uuid: treeUUID,
    });
    const duplicate = await data.checkForExistingTree(treeUUID);
    expect(duplicate).to.not.be.null;
  });

  it('detects a non-duplicate tree', async () => {
    const treeUUID = uuid();
    const duplicate = await data.checkForExistingTree(treeUUID);
    expect(duplicate).to.be.null;
  });

  it('inserts attributes', async () => {
    const attributes = await data.insertTreeAttributes(1, [
      {
        key: 'color',
        value: 'red',
      },
      {
        key: 'deviceid',
        value: 'test',
      },
    ]);
    expect(attributes).to.not.be.null;
    expect(attributes.length).to.equal(2);
    expect(attributes[1].key).to.equal('deviceid');
    expect(attributes[1].value).to.equal('test');
  });

  it('creates a tree with attributes', async () => {
    const tree = await data.createTree(1, 1, {
      lat: 80,
      lon: 120,
      gps_accuracy: 1,
      note: 'my note',
      timestamp: 1536367800,
      image_url: 'http://www.myimage.org/',
      sequence_id: 1,
      attributes: [
        {
          key: 'color',
          value: 'red',
        },
        {
          key: 'deviceid',
          value: 'test',
        },
      ],
    });
    expect(tree).to.not.be.null;
    expect(tree.attributes.length).to.equal(2);
    expect(tree.attributes[1].id).is.not.null;
    expect(tree.attributes[1].key).to.equal('deviceid');
    expect(tree.attributes[1].value).to.equal('test');
  });

  it('creates a device', async () => {
    const dev = await data.upsertDevice({
      app_version: '1',
      app_build: 1,
      manufacturer: 'manufacturer',
      brand: 'brand',
      model: 'model',
      hardware: 'hardware',
      device: 'device',
      serial: 'serial',
    });
    expect(dev).to.not.be.null;
  });
});


helper.js
const rp = require('request-promise-native');
const convertStringToUuid = require('uuid-by-string');
const config = require('./config/config');

const processPlanter = async (planterObject, res, data) => {
  const planter = await data.findOrCreateUser(
    planterObject.planter_identifier,
    planterObject.first_name,
    planterObject.last_name,
    planterObject.organization,
  );
  const body = { ...planterObject };
  body.phone = planter.phone;
  body.email = planter.email;
  await data.createPlanterRegistration(
    planter.id,
    planterObject.device_identifier,
    body,
  );
  console.log(`processed planter ${planter.id}`);
  res.status(200).json({});
};

const processDevice = async (deviceObject, res, data) => {
  // console.log('/device');
  const device = await data.upsertDevice(deviceObject);
  // console.log('upserted');
  res.status(200).json({ device });
  console.log('/device done');
};

const processCapture = async (captureObject, res, data) => {
  console.log('/capture');
  const { version, planter_identifier, device_identifier, session_id } =
    captureObject;
  delete captureObject.version;

  let user;
  if (version === 1) {
    user = await data.findUser(planter_identifier);
    if (user == null) {
      res.status(404).json({
        error: `planter not found ${planter_identifier}`,
      });
      return;
    }
  }
  console.log('check for existing');
  let duplicate = null;
  if (
    captureObject.uuid !== null &&
    captureObject.uuid !== undefined &&
    captureObject.uuid !== ''
  ) {
    duplicate = await data.checkForExistingTree(captureObject.uuid.toLowerCase());
  }
  if (duplicate !== null) {
    console.log('existing');
    res.status(200).json({ duplicate });
  } else if (config.useFieldDataService === 'true') {
    console.log('prepare to use field data service');
    // translate to field-data capture payload
    const tree = { ...captureObject };
    const { attributes } = tree;

    const absStepCountIndex = attributes.findIndex(
      (a) => a.key === 'abs_step_count',
    );
    let abs_step_count;

    if (absStepCountIndex !== -1) {
      const absStepCountArray = attributes.splice(absStepCountIndex, 1);
      [abs_step_count] = absStepCountArray;
    }

    const deltaStepCountIndex = attributes.findIndex(
      (a) => a.key === 'delta_step_count',
    );
    let delta_step_count;

    if (deltaStepCountIndex !== -1) {
      const deltaStepCountArray = attributes.splice(deltaStepCountIndex, 1);
      [delta_step_count] = deltaStepCountArray;
    }

    const rotationMatrixIndex = attributes.findIndex(
      (a) => a.key === 'rotation_matrix',
    );
    let rotation_matrix;

    if (rotationMatrixIndex !== -1) {
      const rotationMatrixArray = attributes.splice(rotationMatrixIndex, 1);
      [rotation_matrix] = rotationMatrixArray;
    }

    const capture = {
      id: tree.uuid,
      image_url: tree.image_url,
      session_id:
        session_id ||
        convertStringToUuid(device_identifier + planter_identifier),
      lat: tree.lat,
      lon: tree.lon,
      gps_accuracy: tree.gps_accuracy,
      abs_step_count: abs_step_count?.value ?? 0,
      delta_step_count: delta_step_count?.value ?? 0,
      rotation_matrix: rotation_matrix?.value ?? [],
      note: tree.note,
      extra_attributes: attributes,
      captured_at: new Date(tree.timestamp * 1000).toISOString(),
    };
    const options = {
      body: capture,
      json: true, // Automatically stringifies the body to JSON
    };
    console.log('contacting field data service');
    console.log(capture);
    const fieldCapture = await rp.post(
      `${config.fieldDataURL}raw-captures`,
      options,
    );
    res.status(201).json({ fieldCapture });
  } else {
    console.log('Storing tree');
    const tree = await data.createTree(
      user.id,
      device_identifier,
      captureObject,
    );
    console.log(`created tree ${tree.uuid}`);
    res.status(201).json({ tree });
  }
};

module.exports = {
  processPlanter,
  processDevice,
  processCapture,
};


src/data.js
class Data {
  constructor(pool) {
    this.pool = pool;
  }

  async checkForExistingTree(uuid) {
    const getQuery = {
      text: `SELECT *
      FROM trees
      WHERE uuid = $1`,
      values: [uuid],
    };
    const rval = await this.pool.query(getQuery).catch((err) => {
      console.log(`ERROR: FAILED TO GET CHECK UUID ON TREES ${err}`);
      throw err;
    });
    if (rval.rows.length > 0) {
      return rval.rows[0];
    } else {
      return null;
    }
  }

  async createTree(planterId, deviceIdentifier, body) {
    let lat = body.lat;
    let lon = body.lon;
    let gpsAccuracy = Math.floor(body.gps_accuracy);
    let timestamp = body.timestamp;
    let imageUrl = body.image_url; // first image
    let planterPhotoUrl = body.planter_photo_url ? body.planter_photo_url : '';
    let planterIdentifier = body.planter_identifier
      ? body.planter_identifier
      : '';
    let note = body.note ? body.note : '';
    let uuid = body.uuid; // This is required
    let imagesJson = body.images ? body.images : {};
    let domainSpecificDataJson = body.domain_specific_data
      ? body.domain_specific_data
      : {};

    const geometry = 'POINT( ' + lon + ' ' + lat + ')';
    const insertQuery = {
      text: `INSERT INTO
      trees(planter_id,
        lat,
        lon,
        gps_accuracy,
        time_created,
        time_updated,
        image_url,
        estimated_geometric_location,
        planter_photo_url,
        planter_identifier,
        device_identifier,
        note,
        uuid,
        images,
        domain_specific_data
      )
      VALUES($1, $2, $3, $4, to_timestamp($5), to_timestamp($5), $6, ST_PointFromText($7, 4326), $8, $9, $10, $11, $12, $13, $14 ) RETURNING *`,
      values: [
        planterId,
        lat,
        lon,
        gpsAccuracy,
        timestamp,
        imageUrl,
        geometry,
        planterPhotoUrl,
        planterIdentifier,
        deviceIdentifier,
        note,
        uuid,
        JSON.stringify(imagesJson),
        JSON.stringify(domainSpecificDataJson),
      ],
    };

    const rval = await this.pool.query(insertQuery).catch((err) => {
      console.log(`ERROR: FAILED TO CREATE TREE ${err}`);
      console.log(insertQuery);
      throw err;
    });
    const tree = rval.rows[0];

    tree.attributes = await this.insertTreeAttributes(tree.id, body.attributes);

    return tree;
  }

  async insertTreeAttributes(id, attributes) {
    var results = [];
    if (attributes) {
      for (let attribute of attributes) {
        const insert = {
          text: `INSERT INTO tree_attributes
          (tree_id, key, value)
          values
          ($1, $2, $3)
          RETURNING *`,
          values: [id, attribute.key, attribute.value],
        };
        const stored = await this.pool.query(insert);
        results.push(stored.rows[0]);
      }
    }
    return results;
  }

  async trees() {
    let query = {
      text: `SELECT *
      FROM trees
      WHERE active = true`,
    };
    const rval = await this.pool.query(query);
    return rval.rows;
  }

  async treesForUser(planterId) {
    let query = {
      text: `SELECT *
      FROM trees
      WHERE planter_id = $1
      AND active = true`,
      values: [planterId],
    };
    const rval = await this.pool.query(query);
    return rval.rows;
  }

  async createPlanterRegistration(planterId, deviceIdentifier, body) {
    const geom = `SRID=4326;POINT (${body.lon} ${body.lat})`;

    var query = {
      text: 'INSERT INTO planter_registrations ( planter_id, device_identifier, first_name, last_name, organization, phone, email, lat, lon, geom ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10) RETURNING *',
      values: [
        planterId,
        deviceIdentifier,
        body.first_name,
        body.last_name,
        body.organization,
        body.phone,
        body.email,
        body.lat,
        body.lon,
        geom,
      ],
    };
    const rval = await this.pool.query(query);
    return rval.rows[0];
  }

  async findUser(identifier) {
    let query = {
      text: 'SELECT * FROM planter WHERE phone = $1 OR email = $1',
      values: [identifier],
    };
    const data = await this.pool.query(query);
    if (data.rows.length != 0) {
      return data.rows[0];
    } else {
      return null;
    }
  }

  async findOrCreateUser(identifier, first_name, last_name, organization) {
    const user = await this.findUser(identifier);
    if (user == null) {
      var reg = new RegExp('^\\d+$');
      var query2 = null;
      if (reg.test(identifier)) {
        query2 = {
          text: 'INSERT INTO planter (first_name, last_name, organization, phone) VALUES ($1, $2, $3, $4 ) RETURNING *',
          values: [first_name, last_name, organization, identifier],
        };
      } else {
        query2 = {
          text: 'INSERT INTO planter (first_name, last_name, organization, email) VALUES ($1, $2, $3, $4 ) RETURNING *',
          values: [first_name, last_name, organization, identifier],
        };
      }
      const rval = await this.pool.query(query2);
      return rval.rows[0];
    } else {
      return user;
    }
  }

  async upsertDevice(body, callback) {
    let android_id = body['device_identifier'];
    let app_version = body['app_version'];
    let app_build = body['app_build'];
    let manufacturer = body['manufacturer'];
    let brand = body['brand'];
    let model = body['model'];
    let hardware = body['hardware'];
    let device = body['device'];
    let serial = body['serial'];
    let android_release = body['androidRelease'] || body['ios_release'];
    let android_sdk =
      body['androidSdkVersion'] || Math.floor(body['ios_sdk_version']) || null;

    // insert only if one does not exist
    const insert = {
      text: `INSERT INTO devices (android_id) values ($1) ON CONFLICT (android_id) DO NOTHING`,
      values: [android_id],
    };
    console.log('start device insert');
    await this.pool.query(insert);
    console.log('completd device insert');

    const query = {
      text: `UPDATE devices
      SET app_version = ($1),
      app_build = ($2),
      manufacturer = ($3),
      brand = ($4),
      model = ($5),
      hardware = ($6),
      device = ($7),
      serial = ($8),
      android_release = ($9),
      android_sdk = ($10)
      WHERE android_id = ($11)
      RETURNING *`,
      values: [
        app_version,
        app_build,
        manufacturer,
        brand,
        model,
        hardware,
        device,
        serial,
        android_release,
        android_sdk,
        android_id,
      ],
    };
    console.log('start device update');
    const returningDevice = await this.pool.query(query);
    console.log('complete device update');
    return returningDevice.rows[0];
  }
}

module.exports = Data;


index.js
require('dotenv').config();
const express = require('express');
require('express-async-errors');
const bodyParser = require('body-parser');
const { Pool } = require('pg');
const { version } = require('./package.json');
const Data = require('./src/data');
const { processPlanter, processDevice, processCapture } = require('./helper');

const config = require('./config/config');

const pool = new Pool({
  connectionString: config.connectionString,
});

pool.on('connect', (client) => {
  console.log(`connected ${client}`);
});

process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

const data = new Data(pool);

const app = express();
const port = process.env.NODE_PORT || 3005;

app.use(bodyParser.urlencoded({ extended: false })); // parse application/x-www-form-urlencoded
app.use(bodyParser.json()); // parse application/json

// Optional fallthrough error handler
app.use((err, req, res, next) => {
  // The error id is attached to `res.sentry` to be returned
  // and optionally displayed to the user for support.
  res.status(500).send(`Error occurred ${err}`);
  next();
});

app.set('view engine', 'html');

app.post('/planter', async (req, res, next) => {
  try {
    await processPlanter(req.body, res, data);
  } catch (error) {
    next(error);
  }
});

app.post('/tree', async (req, res, next) => {
  try {
    await processCapture({ ...req.body, version: 1 }, res, data);
  } catch (error) {
    next(error);
  }
});

app.put('/device', async (req, res, next) => {
  try {
    await processDevice(req.body, res, data);
  } catch (error) {
    next(error);
  }
});

app.post('/v2/wallet_registrations', async (req, res, next) => {
  try {
    const {
      wallet,
      first_name,
      last_name,
      phone,
      email,
      lat,
      lon,
      user_photo_url,
      registered_at,
    } = req.body;

    await processPlanter(
      {
        planter_identifier: wallet,
        first_name,
        last_name,
        phone,
        email,
        organization: null,
        device_identifier: null,
        lat,
        lon,
      },
      res,
      data,
    );
  } catch (error) {
    next(error);
  }
});

app.post('/v2/device_configurations', async (req, res, next) => {
  try {
    const {
      device_identifier,
      brand,
      model,
      device,
      serial,
      hardware,
      manufacturer,
      app_build,
      app_version,
      os_version,
      sdk_version,
      logged_at,
    } = req.body;

    await processDevice(
      {
        device_identifier,
        app_version,
        app_build,
        manufacturer,
        brand,
        model,
        hardware,
        device,
        serial,
        androidRelease: os_version,
        androidSdkVersion: sdk_version,
      },
      res,
      data,
    );
  } catch (error) {
    next(error);
  }
});

app.post('/v2/captures', async (req, res, next) => {
  try {
    const {
      id,
      session_id,
      image_url,
      lat,
      lon,
      gps_accuracy,
      note,
      abs_step_count,
      delta_step_count,
      rotation_matrix,
      // extra_attributes,
      captured_at,
    } = req.body;

    await processCapture(
      {
        uuid: id,
        image_url,
        lat,
        lon,
        gps_accuracy,
        note,
        session_id,
        attributes: [
          { key: 'abs_step_count', value: abs_step_count },
          { key: 'delta_step_count', value: delta_step_count },
          { key: 'rotation_matrix', value: rotation_matrix },
          // ...extra_attributes,
        ],
        timestamp: Math.floor(new Date(captured_at).getTime() / 1000),
        version: 2,
      },
      res,
      data,
    );
  } catch (error) {
    next(error);
  }
});

app.get('*', async (req, res) => {
  res.status(200).send(version);
});

app.use((err, req, res, next) => {
  res.status(500);
  res.json({ error: err.message });
  next(err);
});

app.listen(port, () => {
  console.log(`listening on port ${port}`);
});

module.exports = app;
